src/support.js

```diff
@@ -302,14 +302,31 @@ function loadDynamicLibrary(lib, flags) {
     dso.module = libModule;
   }
 
+  function cleanupOnError() {
+    var lib_record = LDSO.loadedLibs[handle];
+    delete LDSO.loadedLibNames[lib_record.name];
+    delete LDSO.loadedLibs[handle];
+  }
+
   if (flags.loadAsync) {
-    return getLibModule().then(function(libModule) {
-      moduleLoaded(libModule);
-      return handle;
-    })
+    return getLibModule()
+      .catch(function (e) {
+        cleanupOnError();
+        throw e;
```

Comment:
It doesn't look like this codepath is tested.  I'm also not knowledgeable about promises and how they interact with exceptions.    @kripken ?

Reply:
The exceptions stuff here looks good to me.